File: day8.pl
#!/usr/bin/perl

use strict;
use warnings;

sub part1
{
  my $result = 0;

  if (/.* \| (.*)/)
  {
    foreach my $item (split ' ', $1)
    {
      my $l = length($item);
      if($l == 2 or $l == 3 or $l == 4 or $l == 7)
      {
        $result++;
      }
    }
  }
  return($result);
}

sub matches
{
  my $src = shift;
  my $dst = shift;
  my $x;

  foreach my $x (split //, $src)
  {
    if(index($dst, $x) < 0)
    {
      return(0);
    }
  }
  return(1);
}

sub part2
{
  my @found = ('', '', '', '', '', '', '', '', '', '');
  my @parts = split(' \| ', $_[0]);
  my @feed = split(' ', $parts[0]);
  my @digits = split(' ', $parts[1]);
  my %lookup;
  my $i;
  my $result = 0;

  foreach $i (0..(scalar(@feed) - 1))
  {
    my $tmp = join("", sort(split(//, $feed[$i])));
    $feed[$i] = $tmp;
  }
  foreach $i (0..(scalar(@digits) - 1))
  {
    my $tmp = join("", sort(split(//, $digits[$i])));
    $digits[$i] = $tmp;
  }
  foreach $i (@feed)
  {
    if(length($i) == 2)
    { $found[1] = $i; $lookup{$i} = 1; }
    elsif(length($i) == 3)
    { $found[7] = $i; $lookup{$i} = 7; }
    elsif(length($i) == 4)
    { $found[4] = $i; $lookup{$i} = 4; }
    elsif(length($i) == 7)
    { $found[8] = $i; $lookup{$i} = 8; }
  }

  foreach $i (@feed)
  {
    if(!exists($lookup{$i}))
    {
      if(matches($found[1], $i) and length($i) == 5)
      { $found[3] = $i; $lookup{$i} = 3; }
      elsif(matches($found[4], $i))
      { $found[9] = $i; $lookup{$i} = 9; }
    }
  }

  foreach $i (@feed)
  {
    if(!exists($lookup{$i}))
    {
      if(matches($found[1], $i))
      { $found[0] = $i; $lookup{$i} = 0; }
      elsif(length($i) == 6 and !matches($found[1], $i))
      { $found[6] = $i; $lookup{$i} = 6; }
    }
  }

  foreach $i (@feed)
  {
    if(!exists($lookup{$i}))
    {
      if(matches($i, $found[6]))
      { $found[5] = $i; $lookup{$i} = 5; }
      elsif(!matches($i, $found[6]))
      { $found[2] = $i; $lookup{$i} = 2;}
    }
  }

  foreach $i (@digits)
  {
    $result *= 10;
    $result += $lookup{$i};
  }
  return $result;
}

my $part1 = 0;
my $part2 = 0;

open FIN, '<'.$ARGV[0];
while(<FIN>)
{
  chomp;
  $part1 += part1 $_;
  $part2 += part2 $_;
}
print("Part 1:", $part1, "\n");
print("Part 2:", $part2, "\n");

File: day8.py
#!/usr/bin/env python3

import sys


def parsepart1(line):
    parts = line.strip().split(' | ')
    digits = parts[1].split(' ')
    return len([x for x in digits if len(x) in [2, 3, 4, 7]])


def sortsegments(segs):
    return ''.join(sorted([segs[i] for i in range(len(segs))]))


def matches(src, dst):
    for i in range(len(src)):
        if src[i] not in dst:
            return False
    return True


def parsepart2(line):
    lookup = {}
    found = [''] * 10
    parts = line.strip().split(' | ')
    feed = [sortsegments(x) for x in parts[0].split(' ')]
    digits = [sortsegments(x) for x in parts[1].split(' ')]
    result = 0

    for item in feed:
        if len(item) == 2:
            found[1] = item
        elif len(item) == 3:
            found[7] = item
        elif len(item) == 4:
            lookup[item] = '4'
            found[4] = item
        elif len(item) == 7:
            found[8] = item

    for item in feed:
        if item in found:
            continue
        if matches(found[1], item) and len(item) == 5:
            found[3] = item
        elif matches(found[4], item):
            found[9] = item

    for item in feed:
        if item in found:
            continue
        if matches(found[1], item):
            found[0] = item
        elif len(item) == 6 and not matches(found[1], item):
            found[6] = item

    for item in feed:
        if item in found:
            continue
        if matches(item, found[6]):
            found[5] = item
        elif not matches(item, found[6]):
            found[2] = item

    for item in digits:
        result *= 10
        result += found.index(item)

    return result


def main():
    part1 = 0
    part2 = 0
    with open(sys.argv[1], 'r') as fhan:
        for line in fhan.readlines():
            part1 += parsepart1(line)
            part2 += parsepart2(line)
    print("Part 1: {}".format(part1))
    print("Part 2: {}".format(part2))


if __name__ == '__main__':
    main()
